Add guide pages

aazuspan opened the pull request:
Closes #15 

@grovduck, this is an effort to start fleshing out a user guide. There's more to add (tutorials, datasets, other package comparisons, etc.), but I wanted to start with a hopefully manageable chunk that covers core functionality. I wrote a lot of this a couple months ago and did my best to go through and update everything with the recent changes like arbitrary dimensionality, moving away from "image" terminology, etc, but let me know if you find anything that's become outdated.

Any other feedback on the writing or what we could add/modify/remove to make things clearer is welcome!

grovduck reviewed (changes requested):
@aazuspan, this is a great start of the documentation and does a nice job of covering the key bits of the API.  I have a few questions/suggestions, but nothing too substantial.  As always, this is 90% complete on the first go!

Some general comments:
- On the `Home` page, I think it would be valuable to take the first paragraph from the `Getting Started` page and put it at the top (and probably delete the first bullet point).  I think stating the core functionality of the package up front is useful - for me, it gets a bit lost if it's in the feature list. (Kind of like what we have for `sknnr` in the `What is sknnr?` section)  
- Side comment: What's your plan for releasing this via PyPI?  I notice the install is still pointing to the Github repo.  Any worries about `sklearn-raster` getting taken?!
- For `sknnr`, we no longer have a `CONTRIBUTING.md` page at the project root, just in the `docs` subfolder.  Is one way preferable over the other?  We can change `sknnr` to conform to this current pattern if you like better.
- I didn't carefully review the automated docs since those were coming from docstrings and (presumably) we've reviewed those before, but let me know if I need to give those a closer look.

aazuspan commented on docs/mkdocs.yml:
```diff
@@ -40,16 +47,20 @@ plugins:
             merge_init_into_class: true
             signature_crossrefs: true
+            extra:
+              undoc_members: true
```
You're right, that seems to be strictly a Sphinx configuration option. I'm not quite sure what my intention was (it's been couple months since I made the initial changes), but I don't see any difference when I disable it so I think it must have been a mistake. Good catch!

aazuspan commented on docs/pages/guide/raster_formats.md:
```diff
@@ -0,0 +1,85 @@
+`sklearn-raster` supports applying estimator methods to rasters stored in a variety of formats, referred to collectively as **feature arrays**. Numpy arrays can be used for small rasters that fit easily in memory, while Xarray `Dataset` and `DataArray` are ideal for large datasets that benefit from deferred, parallel computation. Estimator methods will return data in the same format that it is pr ...
```
I'd like to stick with the `xr` alias since that's pretty ubiquitous in usage, but I'll specifically mention "Xarray `xr.Dataset`" here and include the `xarray as xr` import and a type annotation in the code examples below, just to make sure that is explicit somewhere in the page.

grovduck replied:
Those changes look good.

aazuspan commented on docs/pages/guide/metadata.md:
```diff
@@ -0,0 +1,97 @@
+1. Setting target names in variable `long_name` attributes.
+1. Appending time-stamped operations to the `history` attribute to help track provenance. 
+
+By default, estimator methods discard all other attributes to avoid preserving inaccurate metadata, e.g. a `scale_factor` that no longer applies to a predicted target. To preserve all attributes, estimator methods accept a `keep_attrs` parameter.
```
Done!

aazuspan commented on docs/pages/guide/metadata.md:
```diff
@@ -0,0 +1,97 @@
+
+A raster stored as an `xr.DataArray` can only contain one `_FillValue` attribute, so it will be broadcast to all bands, if present.
+
+A raster stored in an `xr.Dataset` can store a single dataset-level `_FillValue` as well as a `_FillValue` for each variable. If *any* band contains a `_FillValue`, NoData values for each band will be based on that attribute, even if some bands do not have that attribute (those bands will be treated as if they do not have a NoData value). If *no* band contains a `_FillValue`, the dataset's `_Fill ...
```
Your confusion is warranted! I definitely over-complicated the inference of `nodata_input` from `_FillValue` attributes, and it's probably better to just fix that now rather than trying to accurately describe the current tangle of logic (I would need to study the code to confidently say how it currently works). 

CF conventions and rioxarray's default behavior both only ever set `_FillValue` at the variable level, i.e. in the global attributes of an `xr.DataArray` or in the attributes of the variables of an `xr.Dataset`. To me, that naturally means that we should only infer `nodata_input` from those locations, and shouldn't extrapolate sideways (e.g. if a `_FillValue` is set for one variable it should not affect the inferred NoData value of other variables regardless of whether they set `_FillValue`) or downwards (e.g. the `_FillValue` of an `xr.Dataset` should be ignored, since it's not intended to be a global attribute.)

I think I'll go ahead and make a PR with that change and rebase this onto it.

grovduck replied:
Sounds good to me.  It does has some complexity that is challenging to deal with.

grovduck commented on docs/pages/guide/getting_started.md:
```diff
@@ -0,0 +1,39 @@
+`sklearn-raster` extends `scikit-learn` and other compatible estimators to work directly with raster data. This allows you to train models with tabular data and predict raster outputs directly while preserving metadata like spatial references, band names, and NoData masks.
+
+## Raster Estimators
```
I like the way you've handled this.  Focus first on `raster` as the most common use-case, but acknowledge the possibility of higher-dimensionality datasets.

aazuspan commented:
@grovduck, I've updated the install instructions for PyPI and rewritten the NoData inference section with the simplified rules. Should be ready for another quick look whenever you have some time :)

grovduck reviewed (approved):
@aazuspan, I made another pass through this and found some minor suggestions.  I'm going to mark this as "approved" as you have the final call on all of these decisions.

aazuspan commented on docs/pages/guide/getting_started.md:
```diff
@@ -0,0 +1,37 @@
+To generate predictions from a raster dataset, instantiate a [`scikit-learn`](https://scikit-learn.org/stable/) estimator, [`wrap`](../api/wrap.md) it into a [`FeatureArrayEstimator`](../api/wrap.md/#sklearn_raster.estimator.FeatureArrayEstimator), then fit[^fit-after-wrap] it with tabular data. The `X` dataset should include predictor features that correspond with your raster[^rasters] bands. Fo ...
```
Fixed!

aazuspan commented on docs/pages/guide/raster_formats.md:
```diff
@@ -0,0 +1,89 @@
+|:-------------:|--------------------------|---------------------|-----------------|--------------------|---------------------|----------------------|
+| `np.ndarray` | ✅ |❌ | ❌ | ❌ | ❌ | ❌ |
+| `xr.DataArray` | ✅ | ✅ | ✅ | ✅ | ✅ | ❌ |
+| `xr.Dataset` | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
```
That was bugging me too - thanks for the suggested fix!

aazuspan commented on docs/pages/guide/metadata.md:
```diff
@@ -0,0 +1,98 @@
+
+When possible, `sklearn-raster` will extract column names from the `X` features and `y` targets during fitting and use these to validate and set raster band names for compatible [raster formats](raster_formats.md).
+
+For example, if an estimator is fit with targets stored in a Pandas dataframe and used to predict an Xarray dataset, the column names will be used as the variable names in the output raster. Other wrapped methods will attempt to set reasonable variable names, like sequential neighbor indexes for `kneighbors`. 
```
Added!

aazuspan commented:
>  In pyproject.toml, can you add {args} to the serve script, e.g.

Good idea. I went ahead and added it to `docs:build` as well, just in case.

aazuspan commented:
Thanks as always for the helpful review, @grovduck!
